File: sdks/local-rp/php/tests/conformance/RevocationsTest.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/../bootstrap.php';

use Csilgen\Generated\ClaimSignature;
use Csilgen\Generated\DomainPublicKey;
use Csilgen\Generated\RevocationCertificate;
use Csilgen\Generated\SignedLocalRpCallbackPayload;
use LinkKeys\LocalRp\LocalRp;
use LinkKeys\LocalRp\Revocation;
use LinkKeys\LocalRp\RevocationError;
use LinkKeys\LocalRp\Time;
use LinkKeys\LocalRp\Wire;

$vectors = loadJson(__DIR__ . '/../../../conformance/revocations.json');

TestKit::test('revocations.constants', function () use ($vectors) {
    TestKit::assertEquals($vectors['tag'], Revocation::TAG);
    TestKit::assertEquals($vectors['quorum'], Revocation::QUORUM);
});

/** @return DomainPublicKey[] */
function revocationsDomainKeys(array $vectors): array
{
    $keys = [];
    foreach ($vectors['domain_keys'] as $k) {
        $keys[] = new DomainPublicKey([
            'key_id' => $k['key_id'],
            'public_key' => hexToBytes($k['public_key_hex']),
            'fingerprint' => $k['fingerprint_hex'],
            'algorithm' => $k['algorithm'],
            'key_usage' => $k['key_usage'],
            'created_at' => $k['created_at'],
            'expires_at' => $k['expires_at'],
            'revoked_at' => $k['revoked_at'],
        ]);
    }
    return $keys;
}

function revocationsCertFromCase(array $case): RevocationCertificate
{
    $c = $case['certificate'];
    $sigs = array_map(fn ($s) => new ClaimSignature([
        'domain' => $s['domain'],
        'signed_by_key_id' => $s['signed_by_key_id'],
        'signature' => hexToBytes($s['signature_hex']),
    ]), $c['signatures']);
    return new RevocationCertificate([
        'target_key_id' => $c['target_key_id'],
        'target_fingerprint' => $c['target_fingerprint'],
        'revoked_at' => $c['revoked_at'],
        'signatures' => $sigs,
    ]);
}

$domainKeys = revocationsDomainKeys($vectors);

foreach ($vectors['certificate_cases'] as $case) {
    TestKit::test('revocations.certificate_cases.' . $case['name'], function () use ($case, $domainKeys) {
        $cert = revocationsCertFromCase($case);
        $count = Revocation::countValidSigners($cert, $domainKeys, $case['verify_domain']);
        TestKit::assertEquals($case['expected_counted_signers'], $count, 'expected_counted_signers');

        if ($case['expected_valid']) {
            TestKit::assertDoesNotThrow(fn () => Revocation::verifyRevocationCertificate($cert, $domainKeys, $case['verify_domain']));
        } else {
            TestKit::assertThrows(fn () => Revocation::verifyRevocationCertificate($cert, $domainKeys, $case['verify_domain']));
        }
    });
}

TestKit::test('revocations.certificate_cbor_roundtrip', function () use ($vectors) {
    $case = null;
    foreach ($vectors['certificate_cases'] as $c) {
        if ($c['name'] === 'valid_quorum_two_siblings') {
            $case = $c;
            break;
        }
    }
    TestKit::assertTrue($case !== null);
    $cert = revocationsCertFromCase($case);
    $bytes = Wire::encodeRevocationCertificate($cert);
    $decoded = Wire::decodeRevocationCertificate($bytes);
    TestKit::assertEquals($cert->targetKeyId, $decoded->targetKeyId);
    TestKit::assertEquals($cert->targetFingerprint, $decoded->targetFingerprint);
    TestKit::assertEquals($cert->revokedAt, $decoded->revokedAt);
    TestKit::assertEquals(count($cert->signatures), count($decoded->signatures));
});

TestKit::test('revocations.application_case', function () use ($vectors, $domainKeys) {
    $app = $vectors['application_case'];
    $env = $app['envelope'];

    $signedPayload = new SignedLocalRpCallbackPayload([
        'payload' => hexToBytes($env['payload_cbor_hex']),
        'signing_key_id' => $env['signing_key_id'],
        'signature' => hexToBytes($env['signature_hex']),
    ]);
    $now = Time::parseRfc3339($app['verify_now']);
    $skew = $app['clock_skew_seconds'];

    // Before revocation: the fetched key list shows the target key with NO
    // revoked_at, so the envelope verifies.
    TestKit::assertDoesNotThrow(function () use ($signedPayload, $domainKeys, $now, $skew) {
        LocalRp::verifyLocalRpCallbackPayload($signedPayload, $domainKeys, $now, $skew);
    }, 'expected_valid_before_revocation');
    TestKit::assertEquals(true, $app['expected_valid_before_revocation']);

    // Apply the referenced certificate (quorum-verified) and mark its
    // target revoked; the SAME envelope must now fail.
    $certCase = null;
    foreach ($vectors['certificate_cases'] as $c) {
        if ($c['name'] === 'valid_quorum_two_siblings') {
            $certCase = $c;
            break;
        }
    }
    $cert = revocationsCertFromCase($certCase);
    Revocation::verifyRevocationCertificate($cert, $domainKeys, $vectors['domain']);

    $afterRevocation = array_map(function (DomainPublicKey $k) use ($cert) {
        if ($k->keyId !== $cert->targetKeyId) {
            return $k;
        }
        return new DomainPublicKey([
            'key_id' => $k->keyId,
            'public_key' => $k->publicKey,
            'fingerprint' => $k->fingerprint,
            'algorithm' => $k->algorithm,
            'key_usage' => $k->keyUsage,
            'created_at' => $k->createdAt,
            'expires_at' => $k->expiresAt,
            'revoked_at' => $cert->revokedAt,
        ]);
    }, $domainKeys);

    TestKit::assertThrows(function () use ($signedPayload, $afterRevocation, $now, $skew) {
        LocalRp::verifyLocalRpCallbackPayload($signedPayload, $afterRevocation, $now, $skew);
    }, 'expected_valid_after_revocation');
    TestKit::assertEquals(false, $app['expected_valid_after_revocation']);
});

exit(TestKit::summary('RevocationsTest') === 0 ? 0 : 1);

File: sdks/local-rp/dart/test/conformance/revocations_test.dart
import 'package:linkkeys_local_rp/linkkeys_local_rp.dart';
import 'package:test/test.dart';

import '../testutil/fixtures.dart';

DomainPublicKey _parseDomainKey(Map<String, dynamic> k) {
  return DomainPublicKey(
    keyId: k['key_id'] as String,
    publicKey: hex(k['public_key_hex'] as String),
    fingerprint: k['fingerprint_hex'] as String,
    algorithm: k['algorithm'] as String,
    keyUsage: k['key_usage'] as String,
    createdAt: k['created_at'] as String,
    expiresAt: k['expires_at'] as String,
    revokedAt: k['revoked_at'] as String?,
  );
}

RevocationCertificate _parseCertificate(Map<String, dynamic> certNode) {
  final signatures = <ClaimSignature>[];
  for (final s
      in (certNode['signatures'] as List).cast<Map<String, dynamic>>()) {
    signatures.add(ClaimSignature(
      domain: s['domain'] as String,
      signedByKeyId: s['signed_by_key_id'] as String,
      signature: hex(s['signature_hex'] as String),
    ));
  }
  return RevocationCertificate(
    targetKeyId: certNode['target_key_id'] as String,
    targetFingerprint: certNode['target_fingerprint'] as String,
    revokedAt: certNode['revoked_at'] as String,
    signatures: signatures,
  );
}

void main() {
  group('revocations.json conformance', () {
    test('certificate cases match expected validity and counted signers',
        () async {
      final d = loadJson('revocations.json') as Map<String, dynamic>;
      expect(d['quorum'], equals(revocationQuorum));
      expect(d['tag'], equals('linkkeys-key-revocation-v1alpha'));

      final domainKeys = (d['domain_keys'] as List)
          .cast<Map<String, dynamic>>()
          .map(_parseDomainKey)
          .toList();

      final cases =
          (d['certificate_cases'] as List).cast<Map<String, dynamic>>();
      expect(cases.length, equals(9));

      for (final c in cases) {
        final name = c['name'] as String;
        final cert =
            _parseCertificate(c['certificate'] as Map<String, dynamic>);
        final verifyDomain = c['verify_domain'] as String;
        final expectedValid = c['expected_valid'] as bool;
        final expectedCounted = c['expected_counted_signers'] as int;

        final counted = await countValidSigners(cert, domainKeys, verifyDomain);
        expect(counted, equals(expectedCounted),
            reason: 'expected_counted_signers mismatch for $name');

        if (expectedValid) {
          await verifyRevocationCertificate(cert, domainKeys, verifyDomain);
        } else {
          await expectLater(
            verifyRevocationCertificate(cert, domainKeys, verifyDomain),
            throwsA(isA<RevocationError>()),
            reason: '$name unexpectedly verified',
          );
        }
      }
    });

    test('application case revocation is applied to the key set', () async {
      final d = loadJson('revocations.json') as Map<String, dynamic>;
      final domain = d['domain'] as String;

      final domainKeys = (d['domain_keys'] as List)
          .cast<Map<String, dynamic>>()
          .map(_parseDomainKey)
          .toList();

      final quorumCase = (d['certificate_cases'] as List)
          .cast<Map<String, dynamic>>()
          .firstWhere((c) => c['name'] == 'valid_quorum_two_siblings');
      final cert =
          _parseCertificate(quorumCase['certificate'] as Map<String, dynamic>);

      final app = d['application_case'] as Map<String, dynamic>;
      final envelope = app['envelope'] as Map<String, dynamic>;
      final signedPayload = SignedLocalRpCallbackPayload(
        payload: hex(envelope['payload_cbor_hex'] as String),
        signingKeyId: envelope['signing_key_id'] as String,
        signature: hex(envelope['signature_hex'] as String),
      );
      final verifyNow = DateTime.parse(app['verify_now'] as String).toUtc();
      final skew = app['clock_skew_seconds'] as int;

      // Before applying the revocation certificate: the fetched key list
      // shows the target key with no revoked_at, so the envelope verifies.
      await LocalRp.verifyLocalRpCallbackPayload(
          signedPayload, domainKeys, verifyNow, skew);

      // Apply the quorum-verified certificate exactly as
      // Complete/RpcClient would: verify it, then drop its target from the
      // trusted key set.
      await verifyRevocationCertificate(cert, domainKeys, domain);
      final afterRevocation =
          domainKeys.where((k) => k.keyId != cert.targetKeyId).toList();

      // After applying: the same envelope must fail signature/key-lookup
      // verification, because its signing key is no longer in the trusted
      // set.
      await expectLater(
        LocalRp.verifyLocalRpCallbackPayload(
            signedPayload, afterRevocation, verifyNow, skew),
        throwsA(isA<LocalRpError>()),
      );
    });
  });
}
